bftengine/src/bftengine/BftExecutionEngine.cpp

```diff
@@ -0,0 +1,406 @@
+  concord::diagnostics::TimeRecorder scoped_timer1(*metrics_.executeWriteRequest);
+  requests_handler_->execute(accumulatedRequests, timestamp, cid, span_wrapper);
+  if (accumulatedRequests.size() == 1) timestamp.request_position++;
+}
+void BftExecutionEngineBase::loadTime(SeqNum sn) {
+  if (config_.timeServiceEnabled) {
+    timestamps[sn].time_since_epoch = time_service_manager_->compareAndUpdate(timestamps[sn].time_since_epoch);
+    LOG_INFO(getLogger(),
+             "Timestamp to be provided to the execution: " << timestamps[sn].time_since_epoch.count() << "ms");
+  }
+}
+
+using namespace concord::diagnostics;
+
+class BlockAccumulationExecutionEngine : public BftExecutionEngineBase {
+ public:
+  BlockAccumulationExecutionEngine(std::shared_ptr<IRequestsHandler> requests_handler,
+                                   std::shared_ptr<ClientsManager> client_manager,
+                                   std::shared_ptr<ReplicasInfo> reps_info,
+                                   std::shared_ptr<PersistentStorage> ps,
+                                   std::shared_ptr<TimeServiceManager<CLOCK_TYPE>> time_service_manager)
+      : BftExecutionEngineBase{requests_handler, client_manager, reps_info, ps, time_service_manager} {}
+  SeqNum addExecutions(const std::vector<PrePrepareMsg*>& ppMsgs) override {
+    SeqNum sn = 0;
+    for (auto ppMsg : ppMsgs) {
+      sn = ppMsg->seqNumber();
+      if (bftEngine::ControlStateManager::instance().getPruningProcessStatus()) return sn;
+      TimeRecorder scoped_timer(*metrics_.executeRequestsInPrePrepareMsg);
+      auto requests_for_execution = collectRequests(*ppMsg);
+      TimeRecorder scoped_timer1(*metrics_.executeRequestsAndSendResponses);
+      loadTime(ppMsg->seqNumber());
+      if (!requests_for_execution.empty()) execute(requests_for_execution, timestamps[ppMsg->seqNumber()]);
+      timestamps.erase(ppMsg->seqNumber());
+      post_exec_handlers_.invokeAll(ppMsg, requests_for_execution);
+    }
+    return sn;
+  }
+
+ private:
+  logging::Logger& getLogger() const {
+    static logging::Logger logger = logging::getLogger("bftEngine.impl.BlockAccumulationExecutionEngine");
+    return logger;
+  }
+};
+
+class SingleRequestExecutionEngine : public BftExecutionEngineBase {
+ public:
+  SingleRequestExecutionEngine(std::shared_ptr<IRequestsHandler> requests_handler,
+                               std::shared_ptr<ClientsManager> client_manager,
+                               std::shared_ptr<ReplicasInfo> reps_info,
+                               std::shared_ptr<PersistentStorage> ps,
+                               std::shared_ptr<TimeServiceManager<CLOCK_TYPE>> time_service_manager)
+      : BftExecutionEngineBase{requests_handler, client_manager, reps_info, ps, time_service_manager} {}
+  SeqNum addExecutions(const std::vector<PrePrepareMsg*>& ppMsgs) override {
+    SeqNum sn = 0;
+    for (auto ppMsg : ppMsgs) {
```

Comment:
Same question - can't it be iterated by const ref?

Reply:
same answer :)

Reply:
can you use auto*?